### ISSUE TITLE
debug: support "console" property in launch.json (RunInTerminal)

### ISSUE THREAD

**Issue by ramya-rao-a** [golang/vscode-go#124]
From https://github.com/microsoft/vscode-go/issues/843

> Please support `"console"` Property in launch.json, which have `"none"`, `"integratedTerminal"`, `"externalTerminal"` three Value like python's, and I can debug my go project at outside console and also support standard input.

This feature request has a lot of upvotes https://github.com/microsoft/vscode-go/issues/843,  please follow the discussion there for more context and user scenarios where this feature would be relevant

**Comment by hyangah** [comment]
I managed to prototype this and let vscode launch the headless delve server in the terminal using DAP `RunInTerminal` request.
I still need to figure out a reliable way to address the race occurred when the debug adapter tries to connect to the headless server. DAP RunInTerminal response sometime arrives before the delve headless server is ready to accept the request. When the connection attempt is made too early, the json-rpc2 seems to be hung silently. Since json-rpc2 is not under active maintenance mode, we may need to consider switching the underlying json rpc package (https://github.com/golang/vscode-go/issues/113)

cc/ @suzmue @polinasok

**Comment by JCzz** [comment]
Any progress on this?

**Comment by polinasok** [comment]
@hyangah

> I managed to prototype this and let vscode launch the headless delve server in the terminal using DAP `RunInTerminal` request. I still need to figure out a reliable way to address the race occurred when the debug adapter tries to connect to the headless server. DAP RunInTerminal response sometime arrives before the delve headless server is ready to accept the request.

What is the sequence of requests/responses here? Something like this?
```
[from client to DA] Launch request with console set to terminal
[to DA from to client] RunInTerminal request with dlv command
   client starts dlv
[from client to DA] RunInTerminal response
   DA tries tries to connect to dlv, but dlv is not yet responsive
```
Can you not introduce a delay like we do in the regular adapter: https://github.com/golang/vscode-go/blob/bc20520b0b96608b50df5ebd6f0be2ab6e66ea2c/src/debugAdapter/goDebug.ts#L697-L709?

**Comment by polinasok** [comment]
The `integratedTerminal` and `externalTerminal` modes are powered by [RunInTerminal request](https://microsoft.github.io/debug-adapter-protocol/specification#Reverse_Requests_RunInTerminal). This is advertised as the mechanism for debugging applications that "read from an interactive terminal and/or need to control their output on the terminal" ([source](https://code.visualstudio.com/updates/v1_5#_additions-to-the-debug-protocol), [source](https://microsoft.github.io/debug-adapter-protocol/overview#launching-and-attaching)).

For a high level summary, here are the use cases mentioned by the users in relationship to the request for `console=*Terminal` feature:
 1. debugging with interactive stdin 
  Examples: https://github.com/microsoft/vscode-go/issues/843#issuecomment-414472448, https://github.com/Microsoft/vscode-go/issues/219
   Stdin cannot be entered into the Debug console. Hence, there is a need for an interactive terminal where it can be entered instead. Plain dlv makes this possible, but the user must start a `dlv --headless` server in one terminal (this is where stdin will go) and a client with `dlv connect` in another terminal (this is where debug commands will go). Hence, the workaround (mentioned at https://github.com/microsoft/vscode-go/issues/843#issuecomment-526888927 and https://github.com/microsoft/vscode-go/issues/843#issuecomment-582558216) is to launch a headless dlv server in a terminal manually and then use remote attach in vscode to connect to it.
 2. debugging interactive applications
 Examples: https://github.com/microsoft/vscode-go/issues/843#issuecomment-582551234, https://github.com/golang/vscode-go/issues/863, https://github.com/microsoft/vscode-go/issues/2151
  A workaround (mentioned in https://github.com/microsoft/vscode-go/issues/843#issuecomment-582551234) involves starting the application in a terminal and then using local attach.
 3. redirecting output
 Examples: https://github.com/microsoft/vscode-go/issues/2204
  In case of launch or local attach, dlv is started from the debug adapter, so its output is mixed with the logging and error messages from the adapter in Debug Console. Similarly to 1, a workaround to separate these is to start dlv manually in a separate terminal and then connect to it with remote attach from vscode. In that case dlv's and launched program output would end up in the terminal and not in Debug Console.

Of course, the workarounds do not offer a great user experience. But in the absence of the console option, the users are asking if we could offer some shortcuts for these workarounds via the extra extension settings and features. In the meantime, dlv added new options that look related `--allow-non-terminal-interactive`, `--tty`, `-r` (https://github.com/go-delve/delve/blob/master/Documentation/usage/dlv_redirect.md).

**Comment by hyangah** [comment]
## Implementation proposal

The [console](https://microsoft.github.io/debug-adapter-protocol/overview#launching-and-attaching) attribute provides users an option to specify how to launch and interact with the debugged program (debugee).
internalConsole (default): the debugee output is shown in the default debug console. (e.g. VS Code Debug Console). The program does not have access to stdin.
integratedTerminal: The debugee will run in a terminal emulator integrated in IDE (e.g.  [VS Code Integrated Terminal](https://code.visualstudio.com/docs/editor/integrated-terminal)) and have access to the usual terminal control codes and stdin/stdout/stderr.
externalTerminal: The debugee runs in a separate console window outside of IDE.

Support for integratedTerminal and externalTerminal are available via DAP’s RunInTerminal request and can address many issues that stemmed from inability to access terminal from debugged programs. See [@polinasok’s summary](https://github.com/golang/vscode-go/issues/124#issuecomment-764447497) for the known issues.

### Overview

Dlv dap command tightly integrates the debug-adapter-as-server functionality and the traditional debugger functionality. For the sake of simplicity, this document will use **"Delve (server)”** to refer to the debug-adapter-as-server functionality, and **“Delve (debugger)”** to refer to the debugger functionality.

When receiving a launch request with the console property, Delve (server) will ask the editor to start the Delve (debugger) using the RunInTerminal request. Then the Editor will start the Delve (debugger) in the specified console. To know exactly when the Delve (debugger) is ready, Delve (adapter) will open a TCP port for rendezvous. Once the Delve (debugger) is ready, it will dial into the rendezvous port. Once they are connected, Delve (server) will forward the cached initialize request and launch request and work as a simple proxy between the Editor and the Delve (debugger). 

![runinterminal](https://user-images.githubusercontent.com/4999471/125970769-a15222ea-1462-46a5-9959-139cfae52422.png)

* RunInTerminal response includes a valid process ID (note: sometimes it's empty :-(). If RunInTerminal response succeeds but the Delve (debugger) fails to connect within a certain time limit, Delve (server) will send SIGINT to the process and terminates the debug session.
* Rendezvous port will be closed soon after Delve (debugger) arrives.
* Implementation-wise, Delve (debugger) shares the same DAP handler logic, but does not run a TCP server that accepts from DAP clients.
* Delve (debugger) only works for this debug session and exits when the debug session ends. Multiple client support doesn't apply.
* Delve (server) forwards the client’s capabilities information because they affect the DAP handler logic.
* Delve (server) will not forward RunInTerminal requests from Delve (debugger). That allows arbitrary code execution in the editor side, and we want to close this attack vector.
* So far, we assumed Delve (server) started with the right environment variable and skipped processing of the `env` attribute. In this proposal, Delve (debugger) runs in a separate terminal, we need to tell Delve (debugger) to use the same environment variables. RunInTerminal accepts ‘env’ parameter, but prepends ‘env’ in the command line. Sending the full list of environment variables as RunInTerminal ‘env’ parameter floods the terminal. Either we need to shrink the environment variable list or make Delve (debugger) recognize the `env` attribute.

### Prior Art

According to DAP Spec, RunInTerminal _“is typically used to launch the debuggee in a terminal provided by the client.”_

How to launch the debugee is an implementation detail.

Many debug adapters maintain a clear separation between the adapter and the debugger. Even when they share a single frontend, often they have a special “launcher” command or script that starts the target program with a debugger attacher, or starts the target program with the runtime debugging functionality enabled. 

A typical launch request handling, that does not involve an integrated/external terminal, starts the target program with the launcher. Then the adapter communicates with the debugger or the runtime. When an integrated/external terminal is needed, those debug adapters just need to ask the editor to run the same launch command.

Here are some examples: 
(DISCLAIMER: I gathered the following info by skimming through their source, so it’s possible that some details here may not be quite right. Let me know if need correction)

#### Debugpy

[debugpy](https://github.com/microsoft/debugpy) is a debugger that implements DAP. Their implementation is the primary inspiration of this proposal, so it shares much commonality. When debugpy server receives a launch request, it launches the debugee using [the launcher program](https://github.com/microsoft/debugpy/tree/b07f372d39dd1062109a8ceffcde6dd2b28e3d12/src/debugpy/launcher)which [connects back to the debug adapter](https://github.com/microsoft/debugpy/blob/b07f372d39dd1062109a8ceffcde6dd2b28e3d12/src/debugpy/launcher/__main__.py#L44-L57) at start (similar to the rendezvous port in this proposal) and adds DAP handler. And the debugpy server works as a proxy and delegates request handling to the handler started by the launcher. 

When an integrated/external terminal is needed, the debug adapter sends the launch command to the editor instead of running it directly. A typical launch command sent with RunInTerminal looks like this:

```
$ /usr/bin/env /usr/local/bin/python3 \
   /python-extension/pythonFiles/lib/python/debugpy/launcher 57491 \
   -- /Users/gopher/project/helloworld.py 
```

To enable debugging with root privileges, the launcher will start with `sudo`.

Reference: Debug Adapter [code]( https://github.com/microsoft/debugpy/blob/2ae7dc96ecc4ca73b119c2056284d203afa10afd/src/debugpy/adapter/launchers.py)

The old python debug adapter [ptvsd](https://github.com/microsoft/ptvsd) is also implemented in the same way. See the implementation https://github.com/microsoft/ptvsd/tree/master/src/ptvsd.

#### MIEngine (c++ debug adapter, open-source)

[MIEngine](https://github.com/microsoft/MIEngine) is an open-sourced version of C++ debug adapter VS Code is using. The debug adapter and the debugger are separate, and the debug adapter launches the debugger and debugee (platform-dependent). Unlike python debug adapters that communicate mostly over a TCP socket with the launcher, it sets up pipes (or named pipes for Windows) for communication with the debugger when launching the debugger using RunInTerminal.

For debugging with root privileges, it launches the debugger with the sudo command.
Reference: RunInTerminalTransport.cs [code](
https://github.com/microsoft/MIEngine/blob/8456c49b4731971b885c2909f9feaac4bdcd494d/src/MICore/Transports/RunInTerminalTransport.cs)

#### LLDB Debug Adapter (codelldb)
This is [a debug adapter for LLDB](https://github.com/vadimcn/vscode-lldb), so the roles of debug adapter and the debugger are cleanly separated. It’s possible to run a LLDB command using RunInTerminal, but they took a different path. The debug adapter (codelldb) requests the editor to run [the codelldb “in a terminal agent mode”](https://github.com/vadimcn/vscode-lldb/blob/cf886f402476517379f5d9b43c89639a15e3b793/adapter/loader/terminal_agent.rs). The terminal agent mode codelldb provides access to the tty. The adapter waits on a TCP socket waiting for the terminal agent mode codelldb dials in, provides the tty info, and connects the tty with LLDB/debugee’s stdio streams. This allows the debug adapter to manage the debugee.

Running LLDB as root is not directly supported. Users either have to disable the security feature, or use the remote debugging (start lldb-server in platform mode and connect to it).

### Alternatives considered

#### Editor-dependent solution
In VS Code, we can create our own terminal and start `dlv dap` from the terminal by sending the command using VS Code Terminal API before starting a debug session using  [DebugAdapterDescriptorFactory.createDebugAdapterDescriptor API](https://code.visualstudio.com/api/references/vscode-api#DebugAdapterDescriptorFactory). 

(+): Change in Delve is unnecessary.
(-): This is an editor specific solution, and other DAP clients have to implement their own hacks.
(-): We cannot take advantage of VS Code team’s engineering effort.
(-): With the Terminal API, we can send the command text and there is no feedback. We need a separate mechanism to detect when `dlv dap` server is actually up and running. Options:
  * Create a pipe (platform-specific solution), pass it to –log-dest, and watch the stream.
  * Keep polling to see the process appear as the child of the shell process and cross fingers.
  * Change Delve to dial back or signal us explicitly. 

#### Launching only the debugee in the terminal and attaching

We considered launching the target program and attach to the program using the PID information RunInTerminal response may include. https://github.com/microsoft/vscode/issues/61640#issuecomment-432696354

(+): Only the exact target program command appears in the terminal, so it may look cleaner.
(-): I don’t know an easy way to start a process in suspended mode. It may be possible to write a separate special launcher to achieve the job (maybe reuse some of Delve’s pkg/proc or service/debugger code), that needs work too.
(-): The pid field of RunInTerminal response may not be available. See  https://github.com/microsoft/vscode/issues/61640#issuecomment-432696354
(-): Delve dap still needs changes to handle a new mode in its teardown logic - it looks like attach mode internally, but it is a launch and we need cleanup. This may not be too difficult though.
(-): Support for debugging with root privileges will be tricky. To attach to a program that runs as a root, either we need to run delve as a root or ask users to remove all security features.

#### Launching a tty agent like codelldb

We considered writing our own terminal agent and connecting the Terminal’s tty with debugged program’s stdio streams. I am not sure how much work is necessary to implement cross-platform solutions.

(+): Looks cleaner from the terminal. Log or error messages from Delve will not pollute the terminal.
(+): There will be only one dlv instance, which may look simpler.
(-): Support for debugging with root privileges needs a different route. (codelldb doesn’t support directly).

### TBD

* Log forwarding; ideally we want to minimize polluting the terminal with delve's log messages. (--log, --log-output, --log-dest...)
* Launching Delve (debugger) with more specific commands (dlv debug/test/exec) formulated from the launch request. That will shrink the vulnerable window further.

**Comment by ycchiranjeevi** [comment]
Just Curios, the dlv-dap supports stdin?

If not please provide launch.json and task.json configuration to work with inputing values in separate proccess.

**Comment by asilverman** [comment]
I got something to work, adding to this issue in case it might help:
//launch.json
```json
{
    // Use IntelliSense to learn about possible attributes.
    // Hover to view descriptions of existing attributes.
    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
    "version": "0.2.0",
    "configurations": [
        {
            "preLaunchTask": "start & connect to dlv-dap",
            "name": "Connect to server",
            "type": "go",
            "request": "launch",
            "mode": "debug",
            "port": 2345,
            "host": "127.0.0.1",
            "program": "${fileDirname}",
            "env": {},
            "args": []
        },
    ]
}
```

//tasks.json
```json
{
    // See https://go.microsoft.com/fwlink/?LinkId=733558
    // for the documentation about the tasks.json format
    "version": "2.0.0",
    "tasks": [
        {
            "label": "start dlv-dap",
            "type": "shell",
            "command": "~/go/bin/dlv-dap dap --headless --listen=:2345 --log --api-version=2",
            "problemMatcher": ["$go"],
            "group": {
                "kind": "build",
                "isDefault": true
            },  
            "isBackground": true,     
        }
    ]
}
```

**Comment by hickersonj** [comment]
@asilverman this works for me as far as auto launching dlv dap, but I did have to update the tasks.json label value to match the preLaunchTask value.  

I get a warning about no problem matcher being defined.  Do you have a separate $go problemMatcher? Not sure how to quell that warning or if it matters.
![Screen Shot 2021-12-23 at 1 13 56 PM](https://user-images.githubusercontent.com/11929405/147291960-d45370a9-1ccb-4c94-9bc2-07c7a8c642d2.png)

**Comment by hyangah** [comment]
The golang.go-nightly released today 2022.1.505 allows users to set `console` property.
I'd say this is a still experimental feature.

* uninstall golang.go, install golang.go-nightly and reload window.
* update `dlv` to the latest version

Feedback is welcome.

**Comment by qmuntal** [comment]
> The golang.go-nightly released today 2022.1.505 allows users to set console property.
I'd say this is a still experimental feature.

> uninstall golang.go, install golang.go-nightly and reload window.
update dlv to the latest version
Feedback is welcome.

@hyangah Great to see progress on this feature! Here is my feedback:

I'm trying to debug using the following config:

==launch.json==
```json
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Launch Package",
            "type": "go",
            "request": "launch",
            "mode": "auto",
            "program": "${fileDirname}",
            "console": "externalTerminal"
        }
    ]
}
```

==settings.json==
```json
{
"terminal.external.windowsExec": "cmd.exe /k \"C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\BuildTools\\Common7\\Tools\\VsDevCmd.bat\""
}
```

When launching a debug session the external console appears on the screen but it does not parse the arguments correctly:
```bash
'C:\Program' is not recognized as an internal or external command,
operable program or batch file.
Press any key to continue . . .
```

On the other hand, it works well if I use the `integratedTerminal`:

==launch.json==
```json
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Launch Package",
            "type": "go",
            "request": "launch",
            "mode": "auto",
            "program": "${fileDirname}",
            "console": "integratedTerminal"
        }
    ]
}
```

==settings.json==
```json
{
"terminal.integrated.profiles.windows": {
        "msvc": {
            "path":  ["cmd.exe"],
            "args": [
                "/k",
                "C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\BuildTools\\Common7\\Tools\\VsDevCmd.bat"
            ]
        }
    },
    "terminal.integrated.defaultProfile.windows": "msvc",
}
```

**Comment by hyangah** [comment]
@qmuntal Many thanks for the feedback. The error message indicates the vscode's request to process the arg and run the batch failed. Even the command requested using DAP runInTerminal wasn't tried. I think this is the same issue as https://github.com/microsoft/vscode/issues/57452

**Comment by birgersp** [comment]
> The golang.go-nightly released today 2022.1.505 allows users to set `console` property. I'd say this is a still experimental feature.
> 
> * uninstall golang.go, install golang.go-nightly and reload window.
> * update `dlv` to the latest version
> 
> Feedback is welcome.

I'm using this property for a project, and it's a lifesaver. Great stuff!

**Comment by qmuntal** [comment]
> @qmuntal Many thanks for the feedback. The error message indicates the vscode's request to process the arg and run the batch failed. Even the command requested using DAP runInTerminal wasn't tried. I think this is the same issue as [microsoft/vscode#57452](https://github.com/microsoft/vscode/issues/57452)

Its almost the same issue but for the integrated console instead of the external one.

I've created a new issue with this finding: https://github.com/microsoft/vscode/issues/140974

**Comment by gustavomassa** [comment]
I can confirm the console property works fine on Linux (bash) with "integratedTerminal", and I'm able to use "CTRL + C" signal.
`/usr/bin/env GOPATH=/home/gustavomassa/go /home/gustavomassa/go/bin/dlv dap --check-go-version=false --client-addr=:37467`

But it does not work for "externalTerminal"
![image](https://user-images.githubusercontent.com/16235598/150460464-4e3eaaa5-09c0-425a-88d9-2845eabe6e15.png)
![image](https://user-images.githubusercontent.com/16235598/150460339-0375d69f-b94d-4877-b2e9-6c235961183f.png)

**Comment by hyangah** [comment]
@gustavomassa  Thanks for the report. What is the error message the 'external' terminal shows? It looks like `dlv` printed out its help message but the screenshot hides the beginning of the error message.

**Comment by gustavomassa** [comment]
> @gustavomassa Thanks for the report. What is the error message the 'external' terminal shows? It looks like `dlv` printed out its help message but the screenshot hides the beginning of the error message.

@hyangah 
I was trying to find the dlv command on the logs, but I could not find it. The dlv output is not showing the error/command paths :(
![image](https://user-images.githubusercontent.com/16235598/150463533-6aeff2a7-97ed-44ae-b996-95e92c9b97c7.png)

**Comment by hyangah** [comment]
@gustavomassa Is `ps` still showing the corresponding `bash` command? On my Linux I see an entry like 
   `bash -c ''/usr/local/google/home/hakim/go/bin/dlv dap --check-go-version=false --client-addr=:40481 ; echo; read -p "Press any key to continue...`. (I couldn't reproduce this so I guess something in your environment tickled a bug in VSCode's runInTerminal implementation.)

**Comment by gustavomassa** [comment]
@hyangah I don't know if you changed anything on the extension/code, but it's working fine now
![image](https://user-images.githubusercontent.com/16235598/151095779-555ec80c-a1fe-4bd0-b8de-5eac40f1f8da.png)

One thing to notice is that the terminal is automatically closed when the program execution ends, can we change this behavior to wait for user input before closing the terminal? For example, if we have an error that terminates the program, the terminal is closed before we can read the error

**Comment by hyangah** [comment]
This is done. The only remaining task is probably an entry in FAQ.


@gustavomassa thanks for following up. 

> can we change this behavior to wait for user input before closing the terminal? 

On my Linux, the last command is `read -p "Press any key to continue...."` which waits for user input before closing the terminal. From your https://github.com/golang/vscode-go/issues/124#issuecomment-1018135663 I see the same. If it doesn't work this time, I recommend filing an issue to vscode team with the details about the platform.

**Comment by hyangah** [comment]
Please find the instruction in https://github.com/golang/vscode-go/wiki/debugging#handling-stdin
